File: Battleship.Shared/Ship.cs
namespace Battleship.Shared;
public class Ship
{
    public ShipType Type { get; }
    public Coordinate Start { get; }
    public Orientation Orientation { get; }
    public int Length => Type.GetSize();
    public List<Coordinate> OccupiedCoordinates { get; }
    public HashSet<Coordinate> Hits { get; } = new();
    public Ship(ShipType type, Coordinate start, Orientation orientation)
    {
        Type = type;
        Start = start;
        Orientation = orientation;
        OccupiedCoordinates = CalculateOccupiedCoordinates();
    }
    private List<Coordinate> CalculateOccupiedCoordinates()
    {
        var coordinates = new List<Coordinate>();
        for (int i = 0; i < Length; i++)
        {
            var coord = Orientation == Orientation.Horizontal
                ? new Coordinate(Start.X + i, Start.Y)
                : new Coordinate(Start.X, Start.Y + i);
            coordinates.Add(coord);
        }
        return coordinates;
    }
    public bool IsSunk => Hits.Count == Length;
    public ShotResult FireShot(Coordinate target)
    {
        if (!OccupiedCoordinates.Contains(target))
        {
            return ShotResult.Miss;
        }
        Hits.Add(target);
        return IsSunk ? ShotResult.Sunk : ShotResult.Hit;
    }
}

File: battleships-go/pkg/models/ship.go
package models

type Ship struct {
	Type                ShipType
	Start               Coordinate
	Orientation         Orientation
	OccupiedCoordinates []Coordinate
	Hits                map[Coordinate]struct{}
}

func NewShip(shipType ShipType, start Coordinate, orientation Orientation) *Ship {
	s := &Ship{
		Type:        shipType,
		Start:       start,
		Orientation: orientation,
		Hits:        make(map[Coordinate]struct{}),
	}
	s.OccupiedCoordinates = s.calculateOccupiedCoordinates()

	return s
}
func (s *Ship) Length() int {
	return s.Type.Size()
}

func (s *Ship) calculateOccupiedCoordinates() []Coordinate {
	length := s.Length()
	coords := make([]Coordinate, 0, length)

	for i := 0; i < length; i++ {
		var coord Coordinate
		if s.Orientation == Horizontal {
			coord = Coordinate{X: s.Start.X + i, Y: s.Start.Y}
		} else {
			coord = Coordinate{X: s.Start.X, Y: s.Start.Y + i}

		}

		coords = append(coords, coord)
	}
	return coords
}

func (s *Ship) IsSunk() bool {
	return len(s.Hits) == s.Length()
}

func (s *Ship) ContainsCoordinate(target Coordinate) bool {
	for _, coord := range s.OccupiedCoordinates {
		if coord == target {
			return true
		}
	}
	return false
}

func (s *Ship) FireShot(target Coordinate) ShotResult {
	if !s.ContainsCoordinate(target) {
		return Miss
	}
	s.Hits[target] = struct{}{}

	if s.IsSunk() {
		return Sunk
	}

	return Hit
}
